lambda/matching.ml

```diff
@@ -1178,166 +1194,76 @@ let rec split_or argo cls args def =
-   submatrix for each syntactically-distinct constructor, with a threading
-   of exits such that each submatrix falls back to the potentially-compatible
-   submatrices below it.
-*)
-and split_naive cls args def k =
-  let rec split_exc cstr0 rev_yes = function
-    | [] ->
-        let yes = List.rev rev_yes in
-        ( { me = Pm { cases = yes; args; default = def };
-            matrix = as_matrix yes;
-            top_default = def
-          },
-          k )
-    | ((p :: _, _) as cl) :: rem ->
-        if group_constructor p then
-          let cstr = pat_as_constr p in
-          if cstr = cstr0 then
-            split_exc cstr0 (cl :: rev_yes) rem
-          else
-            let yes = List.rev rev_yes in
-            let { me = next; matrix; top_default = def }, nexts =
-              split_exc cstr [ cl ] rem
-            in
-            let idef = next_raise_count () in
-            let def = Default_environment.cons matrix idef def in
-            ( { me = Pm { cases = yes; args; default = def };
-                matrix = as_matrix yes;
-                top_default = def
-              },
-              (idef, next) :: nexts )
-        else
-          let yes = List.rev rev_yes in
-          let { me = next; matrix; top_default = def }, nexts =
-            split_noexc [ cl ] rem
-          in
-          let idef = next_raise_count () in
-          let def = Default_environment.cons matrix idef def in
-          ( { me = Pm { cases = yes; args; default = def };
-              matrix = as_matrix yes;
-              top_default = def
-            },
-            (idef, next) :: nexts )
-    | _ -> assert false
-  and split_noexc rev_yes = function
+and split_no_or cls args def k =
+  (* We split the remaining clauses in as few pms as possible while maintaining
+     the property stated earlier (cf. {1. Precompilation}), i.e. for
+     any pm in the result, it is possible to decide for any two patterns
+     on the first column whether their heads are equal or not.
+
+     This generally means that we'll have two kinds of pms: ones where the first
+     column is made of variables only, and ones where the head is actually a
+     discriminating pattern.
+
+     There is some subtlety regarding the handling of extension constructors
+     (where it is not always possible to syntactically decide whether two
+     different heads match different values), but this is handled by the
+     [can_group] function. *)
+  let rec split cls =
+    let discr = what_is_first_case cls in
+    collect discr [] [] cls
+  and collect group_discr rev_yes rev_no = function
     | ([], _) :: _ -> assert false
-    | [] -> precompile_var args (List.rev rev_yes) def k
+    | [ ((ps, _) as cl) ] when List.for_all group_var ps && rev_yes <> [] ->
+        (* This enables an extra division in some frequent cases:
+               last row is made of variables only
+           Cf the first part of testsuite/tests/basic/patmatch_split_no_or.ml *)
```

Comment:
This comment seems a bit at odd with the previous one, since it talks positively about adding an extra split (in the case where a non-zero sequences of row with a variable head is ended by an all-variable row) whereas the rest of the code try to reduce the number of splits. Am I missing something, or is the interest of the split more about having a nice catch-all exit point?

Reply:
I will update the comment. In the meantime, here is a (perhaps a bit rough) description of why we think this particular split is good to have:
> If we did not split here, then the whole group would have this catch-all row in its default-environment matrix, and it would never disappear due to further specialization. With the split here, then the first part of the group may be specialized away at some point, and the one-row group that we are creating would remain alone in the default environment; pattern matchings that fail earlier on would jump to it directly, which is more efficient than going through the first part of the matrix first.
This is not very clear, but the gist of it is: we're likely to avoid superfluous tests if we can jump directly to this last clause.

Reply:
Note: the optimization has been there in the matching code for ages and we have been trying to understand why it is, indeed, an optimization. The explanations above are arguments that @trefis and myself came up with (independently) to justify the optimization -- after the fact. Even if the explanation is not very good (maybe we just don't understand), we want to keep the optimization behavior unchanged to avoid any regression.

Reply:
A new comment was proposed by @trefis in https://github.com/ocaml/ocaml/pull/8869/commits/3caa8d58182c5f36b7cfbd6d79f08836f9fd5536.